cpp/src/io/utilities/datasource.cpp

```diff
@@ -40,12 +42,492 @@ namespace cudf {
+}
+
+/**
+ * @brief Helper macro for wrapping a check_read() call in a CUDF_EXPECTS().
+ *
+ * @param bytes_read Supplies the return value from a read().
+ *
+ * @param read_size Supplies the expected number of bytes to have been read.
+ */
+#define CUDF_EXPECTS_READ_SUCCESS(bytes_read, read_size) \
+  CUDF_EXPECTS(check_read(bytes_read, read_size), "read failed")
+
+/**
+ * @brief Host-based data source that issues standard POSIX file I/O calls.
+ */
+class host_source : public datasource {
+ public:
+  host_source(std::string const& filepath) : datasource(datasource_kind::HOST), _filepath(filepath)
+  {
+    // Open the file, then obtain its size by way of fstat().
+    _fd = open(filepath.c_str(), O_RDONLY);
+    if (_fd < 0) {
+      CUDF_LOG_ERROR("Cannot open file {}: {}: {}", filepath, errno, strerror(errno));
+      CUDF_FAIL("Cannot open file");
+    }
+
+    // File descriptor is valid; now obtain the file size.
+    struct stat statbuf;
+    if (fstat(_fd, &statbuf) < 0) {
+      CUDF_LOG_ERROR("Cannot stat file {}: {}: {}", filepath, errno, strerror(errno));
+      CUDF_FAIL("Cannot stat file");
+    }
+    _size = statbuf.st_size;
+  }
+
+  ~host_source() override
+  {
+    if (_fd >= 0) {
+      if (::close(_fd) < 0) {
+        CUDF_LOG_ERROR("Cannot close file {}: {}: {}", _filepath, errno, strerror(errno));
+      }
+    }
+    _fd = -1;
+  }
+
+  [[nodiscard]] std::unique_ptr<buffer> host_read(size_t offset, size_t size)
+  {
+    // Clamp length to available data
+    auto const read_size = get_read_size(size, offset);
+
+    std::vector<uint8_t> v(read_size);
+    auto const bytes_read = host_read(_fd, offset, read_size, v.data());
+    CUDF_EXPECTS_READ_SUCCESS(bytes_read, read_size);
+    return buffer::create(std::move(v));
+  }
+
+  [[nodiscard]] size_t host_read(size_t offset, size_t size, uint8_t* dst)
+  {
+    // Clamp length to available data
+    auto const read_size  = get_read_size(size, offset);
+    auto const bytes_read = host_read(_fd, offset, read_size, dst);
+    CUDF_EXPECTS_READ_SUCCESS(bytes_read, read_size);
+    return read_size;
+  }
+
+  [[nodiscard]] size_t size() const override { return _size; }
+
+ protected:
+  [[nodiscard]] const std::string& filepath() const { return _filepath; }
+
+  /**
+   * @brief Reads a range of bytes from a file descriptor into the supplied
+   * buffer.
+   *
+   * @param fd Supplies the file descriptor from which to read.
+   * @param offset Supplies the offset within the file to begin reading.
+   * @param read_size Supplies the number of bytes to read.  This should be the
+   * clamped read size value obtained from an earlier call to get_read_size().
+   * @param dst Supplies the buffer into which to read the data.  This buffer
+   * should be at least read_size bytes in length.
+   *
+   * @return The number of bytes read on success.  An exception is thrown on
+   * error.
+   */
+  [[nodiscard]] size_t host_read(int fd, size_t offset, size_t read_size, uint8_t* dst)
```

Comment:
This is file_source without GDS, right? should `file_source` be implemented like this? Look naive compared to these functions.
Another question is whether `file_source` and `host_source` should be merged somehow.

Reply:
The role of the existing `file_source` needs to be reviewed in light of what's being proposed--not only that, but there's not even a way to create `file_source` using the updated `create()` factory routine, so it's technically dead code.

I don't think you can say `host_source` is the same as the current `file_source` minus GDS.  The current `file_source` is a hybrid GDS-kvikio implementation, and benefits from kvikio threadpool I/O if GDS isn't available.  But based on my current understanding, it also issues small I/O (`< device_read_threshold`) via kvikio, which may be farmed out to a threadpool, which doesn't seem ideal.  (Although I say this without any empirical data.)

My proposal would be to retire the existing `file_source`, which attempts to tackle a bunch of things depending on runtime environment, then just have very specific derived implementations that Just Do One Thing, then better logic in the `create()` factory routine (and a default `AUTO`) that picks the most sane option at runtime.